src/assets/TokensController.ts

```diff
@@ -505,66 +647,120 @@ export class TokensController extends BaseController<
+   * @param params.newIgnoredTokens - The new ignored tokens to set for the current network and selected account.
+   * @param params.newDetectedTokens - The new detected tokens to set for the current network and selected account.
    * @returns The updated `allTokens` and `allIgnoredTokens` state.
    */
-  _getNewAllTokensState(newTokens: Token[], newIgnoredTokens: string[]) {
-    const { allTokens, allIgnoredTokens } = this.state;
+  _getNewAllTokensState(params: {
+    newTokens?: Token[];
+    newIgnoredTokens?: string[];
+    newDetectedTokens?: Token[];
+  }) {
+    const { newTokens, newIgnoredTokens, newDetectedTokens } = params;
+    const { allTokens, allIgnoredTokens, allDetectedTokens } = this.state;
     const { chainId, selectedAddress } = this.config;
-    const networkTokens = allTokens[chainId];
-    const networkIgnoredTokens = allIgnoredTokens[chainId];
 
-    const newNetworkTokens = {
-      ...networkTokens,
-      ...{ [selectedAddress]: newTokens },
-    };
-    const newIgnoredNetworkTokens = {
-      ...networkIgnoredTokens,
-      ...{ [selectedAddress]: newIgnoredTokens },
-    };
+    let newAllTokens = allTokens;
+    if (newTokens) {
+      const networkTokens = allTokens[chainId];
+      const newNetworkTokens = {
+        ...networkTokens,
+        ...{ [selectedAddress]: newTokens },
+      };
+      newAllTokens = {
+        ...allTokens,
+        ...{ [chainId]: newNetworkTokens },
+      };
+    }
 
-    const newAllTokens = {
-      ...allTokens,
-      ...{ [chainId]: newNetworkTokens },
-    };
+    let newAllIgnoredTokens = allIgnoredTokens;
+    if (newIgnoredTokens) {
+      const networkIgnoredTokens = allIgnoredTokens[chainId];
+      const newIgnoredNetworkTokens = {
+        ...networkIgnoredTokens,
+        ...{ [selectedAddress]: newIgnoredTokens },
+      };
+      newAllIgnoredTokens = {
+        ...allIgnoredTokens,
+        ...{ [chainId]: newIgnoredNetworkTokens },
+      };
+    }
 
-    const newAllIgnoredTokens = {
-      ...allIgnoredTokens,
-      ...{ [chainId]: newIgnoredNetworkTokens },
-    };
-    return { newAllTokens, newAllIgnoredTokens };
+    let newAllDetectedTokens = allDetectedTokens;
+    if (newDetectedTokens) {
+      const networkDetectedTokens = allDetectedTokens[chainId];
+      const newDetectedNetworkTokens = {
+        ...networkDetectedTokens,
+        ...{ [selectedAddress]: newDetectedTokens },
+      };
+      newAllDetectedTokens = {
+        ...allDetectedTokens,
+        ...{ [chainId]: newDetectedNetworkTokens },
+      };
+    }
+    return { newAllTokens, newAllIgnoredTokens, newAllDetectedTokens };
+  }
+
+  /**
+   * Fetch metadata for a token.
+   *
+   * @param tokenAddress - The address of the token.
+   * @returns The token metadata.
+   */
+  async fetchTokenMetadata(tokenAddress: string): Promise<TokenListToken> {
+    try {
+      const token = await fetchTokenMetadata<TokenListToken>(
+        isHexString(this.config.chainId)
```

Comment:
I'm going to reach out to @danjm to see if we can convert hex to decimal from the extension side so that we can remove this conditional. Niranjana is already converting `chainId` in other places, so it should be anything new. This logic probably shouldn't live in the controllers too since controllers should be platform agnostic.

Reply:
I was hoping we could converge on using `0x`-prefixed hex chain IDs everywhere internally. Decimal string numbers are always ambiguous because there's no prefix to hint what it was meant to be, and we already use hex chain IDs throughout most of the extension. But that's outside the scope of this PR.

Passing in a decimal chain ID on the extension sounds like a good idea for now, to simplify this controller. We should make sure this expectation is clearly documented in the constructor as well, for the `onNetworkStateChange` listener.

Reply:
Removed conditional check. We will be converting from hex to decimal on extension for now. Tracked in extension PR - https://github.com/MetaMask/metamask-extension/pull/14484#issuecomment-1139148752